File: backend/internal/models/transaction.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

type Transaction struct {
	ID          string          `gorm:"type:varchar(21);primary_key" json:"id"`
	SpaceID     uuid.UUID       `gorm:"type:uuid;not null" json:"space_id"`
	Type        string          `gorm:"type:varchar(20);not null;default:'expense'" json:"type"`
	Title       string          `gorm:"type:varchar(100)" json:"title"`
	Date        time.Time       `gorm:"not null;default:NOW()" json:"date"`
	Currency    string          `gorm:"type:varchar(3);not null;default:'TWD'" json:"currency"`
	TotalAmount decimal.Decimal `gorm:"type:decimal(10,2);not null;default:0" json:"total_amount"`
	Note        string          `gorm:"type:text" json:"note"`
	AIStatus    *string         `gorm:"type:varchar(20);column:ai_status" json:"ai_status,omitempty"`
	AIError     string          `gorm:"type:text;column:ai_error" json:"ai_error,omitempty"`
	CreatedAt   time.Time       `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time       `gorm:"autoUpdateTime" json:"updated_at"`

	// Associations
	Space   *Space              `gorm:"foreignKey:SpaceID" json:"space,omitempty"`
	Expense *TransactionExpense `gorm:"foreignKey:TransactionID;constraint:OnDelete:CASCADE" json:"expense,omitempty"`
	Debts   []TransactionDebt   `gorm:"foreignKey:TransactionID;constraint:OnDelete:CASCADE" json:"debts,omitempty"`
	Images  []Image             `gorm:"polymorphic:Entity;polymorphicValue:transaction" json:"images,omitempty"`
}

func (Transaction) TableName() string {
	return "transactions"
}

type TransactionExpense struct {
	ID            uuid.UUID       `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
	TransactionID string          `gorm:"type:varchar(21);not null;uniqueIndex" json:"transaction_id"`
	Category      string          `gorm:"type:varchar(50)" json:"category"`
	ExchangeRate  decimal.Decimal `gorm:"type:decimal(12,6);default:1" json:"exchange_rate"`
	BillingAmount decimal.Decimal `gorm:"type:decimal(10,2);default:0" json:"billing_amount"`
	HandlingFee   decimal.Decimal `gorm:"type:decimal(10,2);default:0" json:"handling_fee"`
	PaymentMethod string          `gorm:"type:varchar(50)" json:"payment_method"`
	LocationURL   string          `gorm:"type:varchar(500);not null;default:''" json:"location_url"`

	// Associations
	Items []TransactionExpenseItem `gorm:"foreignKey:ExpenseID;constraint:OnDelete:CASCADE" json:"items,omitempty"`
}

func (TransactionExpense) TableName() string {
	return "transaction_expenses"
}

type TransactionExpenseItem struct {
	ID        uuid.UUID       `gorm:"type:uuid;primary_key;default:uuid_generate_v4()" json:"id"`
	ExpenseID uuid.UUID       `gorm:"type:uuid;not null" json:"expense_id"`
	Name      string          `gorm:"type:varchar(100);not null" json:"name"`
	UnitPrice decimal.Decimal `gorm:"type:decimal(10,2);not null;default:0" json:"unit_price"`
	Quantity  decimal.Decimal `gorm:"type:decimal(8,2);not null;default:1" json:"quantity"`
	Discount  decimal.Decimal `gorm:"type:decimal(10,2);default:0" json:"discount"`
	Amount    decimal.Decimal `gorm:"type:decimal(10,2);not null;default:0" json:"amount"`
}

func (TransactionExpenseItem) TableName() string {
	return "transaction_expense_items"
}

File: frontend/types/transaction.ts
import type { Image } from './image'

export type TransactionType = 'expense' | 'payment'

export type AiStatus = 'pending' | 'processing' | 'completed' | 'failed'

export interface Transaction {
  id: string
  space_id: string
  type: TransactionType
  title: string
  date: string
  currency: string
  total_amount: string
  note: string
  created_at: string
  updated_at: string
  expense?: TransactionExpense
  debts?: TransactionDebt[]
  images?: Image[]
  /** AI receipt extraction status; undefined when the row never went through AI. */
  ai_status?: AiStatus
  /** User-facing failure message when ai_status === 'failed'. */
  ai_error?: string
}

export interface TransactionExpense {
  id: string
  transaction_id: string
  category: string
  exchange_rate: string
  billing_amount: string
  handling_fee: string
  payment_method: string
  location_url: string
  items?: TransactionExpenseItem[]
}

export interface TransactionExpenseItem {
  id: string
  expense_id: string
  name: string
  unit_price: string
  quantity: string
  discount: string
  amount: string
}

export interface ExpenseTemplateItem {
  name: string
  unit_price: string
  quantity: string
  discount: string
}

export interface ExpenseTemplateDebt {
  payer_name: string
  payee_name: string
  amount: string
  is_spot_paid: boolean
}

export interface ExpenseTemplateData {
  title: string
  category: string
  currency: string
  payment_method: string
  location_url: string
  note: string
  total_amount: string
  items: ExpenseTemplateItem[]
  debts: ExpenseTemplateDebt[]
}

export interface ExpenseTemplate {
  id: string
  space_id: string
  name: string
  data: ExpenseTemplateData
  created_at: string
  updated_at: string
}

export interface TransactionDebt {
  id: string
  transaction_id: string
  payer_name: string
  payee_name: string
  amount: string
  settled_amount: string
  is_spot_paid: boolean
}
